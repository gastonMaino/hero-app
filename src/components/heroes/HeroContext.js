import {createContext} from 'react';

export const HeroeContext = createContext(null);