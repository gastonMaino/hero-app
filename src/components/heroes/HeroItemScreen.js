import React from 'react'

export const HeroItemScreen = ({hero}) => {

    const { publisher, alter_ego, first_appearance, characters, description } = hero;

    return (
        <>
          <li className='list-group__item'>
                        <p className='paragraph'>
                            <span className='list-group__span'>Alter Ego: </span>
                            {alter_ego}
                        </p>
                    </li>

                    {
                        (alter_ego !== characters)
                        &&
                        <li className='list-group__item'>
                            <p className='paragraph'>
                                <span className='list-group__span'>Characters: </span>
                                {characters}
                            </p>
                        </li>
                    }
                    <li className='list-group__item'>
                        <p className='paragraph'>
                            <span className='list-group__span'>
                                First Appearance:   
                            </span>
                            {first_appearance}
                        </p>
                    </li>
                    <li className='list-group__item'>
                        <p className='paragraph'>
                            <span className='list-group__span'>Publisher: </span>
                            {publisher}
                        </p>
                    </li>
                    {
                        (description !== undefined)
                        &&
                        <li className='list-group__item'>
                            <p className='paragraph'>
                                <span className='list-group__span'>Biography: </span> {description}
                            </p>
                        </li>
                    }   
        </>
    )
}
