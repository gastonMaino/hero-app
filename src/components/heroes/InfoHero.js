import React, { useContext } from 'react';
import { Link } from 'react-router-dom';
import { HeroeContext } from './HeroContext';



export const InfoHero = () => {
    const { id, superhero, alter_ego, first_appearance, characters } = useContext(HeroeContext);
    return (
        <>
            <p className='title'>{superhero}</p>
            <p className='paragraph'>{alter_ego}</p>

            {
                (alter_ego !== characters)
                &&
                <p className='paragraph'>{characters}</p>
            }

            <p className='paragraph'>
                <span className='paragraph__span'>
                    {first_appearance}
                </span>
            </p>

            <Link to={`./hero/${id}`} className='button'>
                See more...
                </Link>

        </>
    )
}
