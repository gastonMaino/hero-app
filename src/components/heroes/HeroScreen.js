import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router'
import { getHeroesById } from '../../selectors/getHeroesById';
import { HeroItemScreen } from './HeroItemScreen';

export const HeroScreen = ({history}) => {

    const { heroid } = useParams();

    const hero = useMemo(() => getHeroesById(heroid), [heroid]);

    if (!hero) return <Redirect to='/' />

    const handleReturn = () =>{
        if(history.length <= 2){
            history.push('/')
        }else{

            history.goBack();
        }
    }

    const { superhero } = hero;

    return (
        <article className='article'>
            <div className='container-img'>
                <img 
                src={`../assets/heroes/${heroid}.jpg`} 
                alt={`${superhero}`} 
                />
            </div>
            <div className='container-description'>
                <p className='title'>{superhero}</p>
                <ul className='list-group'>
                   <HeroItemScreen hero={hero} />
                </ul>

                <button className='btn' onClick={handleReturn}>
                    return
                </button>
            </div>
        </article>
    )
}
