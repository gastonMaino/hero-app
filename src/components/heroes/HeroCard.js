import React, { useContext } from 'react';
import { HeroeContext } from './HeroContext';
import { InfoHero } from './InfoHero';

export const HeroCard = () => {

    const {id, superhero} = useContext(HeroeContext)

    return (
        <div className='card'>
            <div className='card-image'>
                <img src={`assets/heroes/${id}.jpg`} alt={`Is a hero ${superhero}`} />
            </div>

            <div className='card-content'>
                <InfoHero />
            </div>
        </div>
    )
}
