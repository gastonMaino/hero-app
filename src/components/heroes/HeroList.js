import React, { useMemo } from 'react';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroCard } from './HeroCard';
import { HeroeContext } from './HeroContext';

export const HeroList = ({ publisher }) => {
    
    const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher]);
    return (
        <section className='parent-card animate__animated animate__fadeIn'>
            {
                heroes.map((hero) => {
                    return (
                        <HeroeContext.Provider key={hero.id} value={hero}>
                            <HeroCard />
                        </HeroeContext.Provider>
                            

                    )
                })
            }
        </section>
    )
}
