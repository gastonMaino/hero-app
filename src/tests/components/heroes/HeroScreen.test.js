import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { HeroScreen } from "../../../components/heroes/HeroScreen"

describe('tests in <HeroScreen />', () => {

    const historyMock = {
        replace: jest.fn(),
        length: 10,
        goBack: jest.fn(),
        push: jest.fn(),
        location: {},
        listen: jest.fn(),
        createHref: jest.fn()
    }


    test('should match to snapshot amd show the Redirect comoponent', () => {
        const wrapper = mount(
            <MemoryRouter>
                <HeroScreen history={historyMock} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true);
    })

    test('should show a hero if is exist', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path='/hero/:heroid' component={HeroScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('article').exists()).toBe(true);
        expect(wrapper.find('.container-img').exists()).toBe(true);
        expect(wrapper.find('HeroItemScreen').exists()).toBe(true);
    })

    test('should call history.push()', () => {
        const historyMock = {
            length: 1,
            goBack: jest.fn(),
            push: jest.fn(),
            location: {},
            listen: jest.fn(),
            createHref: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route
                    path='/hero/:heroid'
                    component={() => <HeroScreen history={historyMock} />}
                />
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')()

        expect(historyMock.push).toHaveBeenCalledWith('/');
        expect(historyMock.goBack).not.toHaveBeenCalled();
    })

    test('should call history.goBack()', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route
                    path='/hero/:heroid'
                    component={() => <HeroScreen history={historyMock} />}
                />
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')()

        expect(historyMock.goBack).toHaveBeenCalled();
        expect(historyMock.push).not.toHaveBeenCalled();
    })

    test('should call Redirect', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider2334']}>
                <Route
                    path='/hero/:heroid'
                    component={() => <HeroScreen history={historyMock} />}
                />
            </MemoryRouter>
        )

        expect(wrapper.text()).toBe('');
    })


})
