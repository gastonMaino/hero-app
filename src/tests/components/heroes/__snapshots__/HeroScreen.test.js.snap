// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in <HeroScreen /> should match to snapshot amd show the Redirect comoponent 1`] = `""`;
